File: python/samples/07_event_monitor.py
#!/usr/bin/env python3
"""Sample 07 — Event listener with type and registry-prefix filtering.

Demonstrates:
  - listener.on_event_type(event_type, callback)
  - listener.on_registry(registry_prefix, callback)
  - Combining multiple callback registrations
  - Async vs sync callbacks

Usage:
    python 07_event_monitor.py [--host HOST] [--port PORT]
                               [--listen-port LISTEN_PORT]
"""

from __future__ import annotations

import argparse
import asyncio
import sys
from pathlib import Path
from datetime import datetime

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

import redfish_sdk
from redfish_sdk import AuthMode, ConnectionConfig, Credentials, RedfishEventListener
from redfish_sdk.services.event_service import RedfishEvent
from redfish_sdk.errors import RedfishConnectionError


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Sample 07 — Event Monitor")
    p.add_argument("--host", default="127.0.0.1")
    p.add_argument("--port", type=int, default=8000)
    p.add_argument("--user", default="admin")
    p.add_argument("--password", default="password")
    p.add_argument("--no-tls-verify", action="store_true")
    p.add_argument("--no-tls", action="store_true", help="Use plain HTTP instead of HTTPS (for simulators without SSL)")
    p.add_argument("--listen-port", type=int, default=9091)
    p.add_argument("--wait", type=float, default=15.0)
    return p.parse_args()


# ── Module-level counters (shared across callbacks) ──────────────────────────
_stats: dict[str, int] = {
    "total": 0,
    "alerts": 0,
    "base_registry": 0,
    "status_change": 0,
}


async def main() -> None:
    args = parse_args()

    creds = Credentials(username=args.user, password=args.password)
    config = ConnectionConfig(verify_tls=not args.no_tls_verify, use_tls=not args.no_tls)

    try:
        ctx = await redfish_sdk.connect_async(
            host=args.host,
            port=args.port,
            credentials=creds,
            auth_mode=AuthMode.SESSION,
            config=config,
        )
    except RedfishConnectionError as exc:
        print(f"[ERROR] {exc}")
        sys.exit(1)

    async with ctx:
        listener = RedfishEventListener(port=args.listen_port)
        listener.use_context(ctx)

        # ── Global catch-all (sync) ──────────────────────────────────────
        def catch_all(event: RedfishEvent) -> None:
            _stats["total"] += 1
            ts = datetime.now().strftime("%H:%M:%S.%f")[:-3]
            print(f"  [ALL      {ts}] {event.event_type} — {event.message_id}")

        listener.on_event(catch_all)

        # ── Alert-type filter (async) ────────────────────────────────────
        async def on_alert(event: RedfishEvent) -> None:
            _stats["alerts"] += 1
            print(f"  [ALERT   ] severity={event.severity!r}  msg={event.message!r}")

        listener.on_event_type("Alert", on_alert)

        # ── StatusChange filter (sync) ───────────────────────────────────
        def on_status_change(event: RedfishEvent) -> None:
            _stats["status_change"] += 1
            print(f"  [STATUS  ] origin={event.origin_of_condition!r}")

        listener.on_event_type("StatusChange", on_status_change)

        # ── "Base" registry filter (async) ───────────────────────────────
        async def on_base_registry(event: RedfishEvent) -> None:
            _stats["base_registry"] += 1
            print(f"  [BASE.REG] {event.message_id} — {event.message}")

        listener.on_registry("Base", on_base_registry)

        # ── Start ────────────────────────────────────────────────────────
        listener.start()
        print(f"✓ Listener active at {listener.listen_url}")
        print(f"  Callbacks: catch_all (all), Alert, StatusChange, Base.*\n")

        dest = f"http://127.0.0.1:{args.listen_port}/events"
        sub_resp = await ctx.events.subscribe_async(
            destination=dest,
            event_types=["Alert", "ResourceUpdated", "StatusChange"],
            context="RSDK-Sample-07",
        )

        sub_uri: str | None = None
        if sub_resp.success:
            sub_uri = sub_resp.headers.get("Location") or sub_resp.body.get("@odata.id")
            print(f"✓ Subscribed — {sub_uri}")
        else:
            print(f"✗ Subscribe HTTP {sub_resp.status_code}")

        # Trigger several test events
        for _ in range(3):
            await ctx.events.submit_test_event_async()
            await asyncio.sleep(0.5)

        print(f"\nWaiting {args.wait}s for events …")
        await asyncio.sleep(args.wait)

        # ── Stats ────────────────────────────────────────────────────────
        print("\n── Event statistics ──────────────────────")
        for key, val in _stats.items():
            print(f"  {key:<16}: {val}")

        # ── Cleanup ──────────────────────────────────────────────────────
        if sub_uri:
            await ctx.events.delete_subscription_async(sub_uri)
        listener.stop()

        print("\n✓ Event monitor sample complete")


if __name__ == "__main__":
    asyncio.run(main())

File: cpp/samples/07_event_monitor.cpp
/**
 * Sample 07 — Event Monitor (filtered callbacks)
 *
 * Demonstrates:
 *   - listener.on_event_type(type, callback)   — EventType-filtered callback
 *   - listener.on_registry(prefix, callback)   — registry-prefix-filtered callback
 *   - Combining multiple callback registrations
 *   - get_ip_stats() / get_buffered_events()
 *
 * Usage:
 *   ./07_event_monitor [host] [port] [listen-port] [wait-seconds]
 */
#include "redfish_sdk/client.hpp"
#include "redfish_sdk/errors.hpp"
#include "redfish_sdk/event_listener.hpp"
#include <atomic>
#include <chrono>
#include <iostream>
#include <thread>

int main(int argc, char* argv[]) {
    std::string host        = (argc > 1) ? argv[1] : "127.0.0.1";
    int         port        = (argc > 2) ? std::stoi(argv[2]) : 8000;
    uint16_t    listen_port = (argc > 3) ? static_cast<uint16_t>(std::stoi(argv[3])) : 9091;
    int         wait_sec    = (argc > 4) ? std::stoi(argv[4]) : 15;

    redfish::Credentials     creds{"admin", "admin"};
    redfish::ConnectionConfig config;
    config.verify_tls = false;  // disable cert verification for simulator (self-signed)
    for (int i = 1; i < argc; ++i)
        if (std::string(argv[i]) == "--no-tls") config.use_tls = false;  // plain HTTP

    try {
        auto ctx = redfish::connect(host, port, creds,
                                    redfish::AuthMode::SESSION, config);

        redfish::RedfishEventListener listener{listen_port};
        listener.use_context(*ctx);

        std::atomic<int> total{0}, alerts{0}, base_msgs{0};

        // ── Global catch-all ──────────────────────────────────────────
        listener.on_event([&](const redfish::RedfishEvent& ev) {
            ++total;
            std::cout << "  [ALL   ] " << ev.event_type
                      << " — " << ev.message_id << "\n";
        });

        // ── EventType-filtered: Alert only ────────────────────────────
        listener.on_event_type("Alert", [&](const redfish::RedfishEvent& ev) {
            ++alerts;
            std::cout << "  [ALERT ] " << ev.message << "\n";
        });

        // ── Registry-prefix-filtered: Base.* messages ─────────────────
        listener.on_registry("Base", [&](const redfish::RedfishEvent& ev) {
            ++base_msgs;
            std::cout << "  [BASE  ] id=" << ev.message_id << "\n";
        });

        // ── OpenBMC registry ──────────────────────────────────────────
        listener.on_registry("OpenBMC", [&](const redfish::RedfishEvent& ev) {
            std::cout << "  [OPENBMC] " << ev.message_id
                      << "  sev=" << ev.severity << "\n";
        });

        listener.start();
        std::cout << "✓ Listener at " << listener.listen_url() << "\n";

        // ── Subscribe ─────────────────────────────────────────────────
        std::string dest = "http://127.0.0.1:" + std::to_string(listen_port)
                           + "/events";
        auto& ev_svc = ctx->events();
        auto  sub    = ev_svc.subscribe(dest,
                                         {"Alert", "ResourceUpdated", "StatusChange"},
                                         {"Base", "OpenBMC"},
                                         {}, {}, "", "RSDK-Sample-07");

        std::string sub_uri;
        if (sub.success) {
            sub_uri = sub.location();
            if (sub_uri.empty() && sub.body.is_object())
                sub_uri = sub.body.value("@odata.id", "");
            std::cout << "  ✓ Subscribed: " << sub_uri << "\n";
        } else {
            std::cout << "  ✗ Subscribe HTTP " << sub.status_code << "\n";
        }

        // Trigger a couple of test events
        ev_svc.submit_test_event();
        std::this_thread::sleep_for(std::chrono::milliseconds(200));
        ev_svc.submit_test_event({{"EventType","StatusChange"},
                                   {"MessageId","Base.1.8.Success"},
                                   {"Message","Status change event"}});

        std::cout << "Waiting " << wait_sec << "s …\n";
        std::this_thread::sleep_for(std::chrono::seconds(wait_sec));

        // ── Summary ───────────────────────────────────────────────────
        std::cout << "\n── Event summary ──\n";
        std::cout << "  Total     : " << total.load()    << "\n";
        std::cout << "  Alerts    : " << alerts.load()   << "\n";
        std::cout << "  Base.*    : " << base_msgs.load()<< "\n";
        std::cout << "  Buffered  : " << listener.get_buffered_events().size()
                  << "\n";
        for (const auto& [ip, cnt] : listener.get_ip_stats())
            std::cout << "  IP " << ip << " → " << cnt << "\n";

        // ── Cleanup ───────────────────────────────────────────────────
        if (!sub_uri.empty()) ev_svc.delete_subscription(sub_uri);
        listener.stop();

        std::cout << "\n✓ Event monitor sample complete\n";
        return 0;
    } catch (const redfish::RedfishError& e) {
        std::cerr << "[ERROR] " << e.what() << "\n";
    }
    return 1;
}
